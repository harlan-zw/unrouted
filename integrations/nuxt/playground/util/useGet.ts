import { $fetch, FetchOptions} from 'ohmyfetch'
import {$get, $put, TypedInternalGetResponse} from '@unrouted/fetch'


export function useGet<R extends string, Default = unknown>(url: R, options = {}) {
  const fetch = useFetch(url, options)
  return {
    data: fetch.data as unknown as Promise<TypedInternalGetResponse<R, Default>>
  }
}

const res = await $put('/api/modules')
let modules = await $get('/api/modules')
