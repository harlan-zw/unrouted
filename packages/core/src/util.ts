import { withBase, withLeadingSlash, withoutTrailingSlash } from 'ufo'
import { promisifyHandle } from 'h3'
import { murmurHash } from 'ohash'
import { defu } from 'defu'
import type { NormaliseRouteFn, RegisterRouteFn } from './types'
import { useUnrouted } from './unrouted'

/**
 * All route paths have a leading slash and no trailing slash.
 */
const normaliseSlashes = (s: string) => withLeadingSlash(withoutTrailingSlash(s))

export const resolveStackPrefix = () => {
  const { groupStack } = useUnrouted()
  let path = ''
  // apply all prefixes from the stack
  for (const prefix of groupStack
    .filter(g => !!g.prefix)
    .map(g => g.prefix)) {
    // ensure consistency, apply prefix, this could be from a group or something
    path = withBase(normaliseSlashes(path), prefix!)
  }
  return path
}

export const resolveStackRouteMeta = () => {
  const { groupStack } = useUnrouted()
  let meta = {}
  // apply all prefixes from the stack
  for (const m of groupStack
    .filter(g => !!g.routeMeta)
    .map(g => g.routeMeta)
    .reverse()) {
    // ensure consistency, apply prefix, this could be from a group or something
    meta = defu(meta, m!)
  }
  return meta
}

/**
 * Create a normalised route from numerous inputs.
 */
export const normaliseRoute: NormaliseRouteFn = (method, path, handle, meta?) => {
  // ensure consistency, apply prefix, this could be from a group or something
  path = withBase(normaliseSlashes(path), resolveStackPrefix())
  if (typeof handle === 'function' && handle.length > 2)
    handle = promisifyHandle(handle)
  if (!Array.isArray(method))
    method = [method]
  meta = defu(meta || {}, resolveStackRouteMeta())
  return {
    id: `_${murmurHash(`${method.join(',')} ${path}`)}`,
    path,
    handle,
    method,
    meta,
    // add functions for chanining @todo
    // where (matches: Record<string, RegExp>) {
    //   meta.parameterMatchRegExps = {
    //     ...meta.parameterMatchRegExps,
    //     ...matches,
    //   }
    // },
    // name(name: string) {
    //   meta.name = name
    // }
  }
}

/**
 * Registers a route with our router.
 *
 * The route is registered asynchronously to allow for plugins to hook into the registration. This means the route won't
 * be immediately available within the Unrouter context.
 */
export const registerRoute: RegisterRouteFn = (method, path, handle, options?) => {
  const route = normaliseRoute(method, path, handle, options)
  const { routes, logger } = useUnrouted()
  if (routes.map(r => r.path).includes(path)) {
    logger.debug('Skipping duplicate route registration')
    return route
  }
  // figure out what code registered this for better error handling and type support
  logger.debug(`Registering route \`${method}\` \`${route.path}\`.`)
  routes.push(route)
  return route
}
