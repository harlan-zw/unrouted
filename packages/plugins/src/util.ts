import { generateTypes, resolveSchema } from 'untyped'

interface GenerateTypesOptions {
  interfaceName?: string
  addExport?: boolean
  addDefaults?: boolean
  defaultDescrption?: string
  indentation?: number
  allowExtraKeys?: boolean
}

export function untypePayload(payload: any, options?: GenerateTypesOptions) {
  if (typeof payload === 'string')
    return `'${payload}'|${typeof payload}`
  if (typeof payload === 'number')
    return `${payload}|${typeof payload}`
  return generateTypes(resolveSchema(payload), options)
}
