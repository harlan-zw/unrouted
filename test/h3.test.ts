import { describe } from 'vitest'
import { bootstrap } from './util'

describe('h3 provider', async () => {
  await bootstrap()
})
