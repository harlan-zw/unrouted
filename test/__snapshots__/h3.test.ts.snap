// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`h3 provider > serving static works 1`] = `
"<html>
<head>
    <meta charset="UTF-8">
    <meta name="viewport"
          content="width=device-width, user-scalable=no, initial-scale=1.0, maximum-scale=1.0, minimum-scale=1.0">
    <meta http-equiv="X-UA-Compatible" content="ie=edge">
    <title>Document</title>
</head>
<body>
    <p>Lorem ipsum dolor sit amet, consectetur adipisicing elit. Architecto asperiores deserunt ducimus eius enim et, facere fugit laborum nemo nostrum officia pariatur possimus recusandae reprehenderit sed tempore totam ullam velit.</p>
</body>
</html>
"
`;
